### ISSUE TITLE
Low CPU utilisation 

### ISSUE THREAD

**Issue by deepkshikha** [AlexeyAB/darknet#667]
I am using two instance one is m5.4xlarge general purpose instance 
![cpuut](https://user-images.githubusercontent.com/13184577/38992159-829c11ea-43fd-11e8-9141-85ad1b7b674c.PNG)

![gpucpu](https://user-images.githubusercontent.com/13184577/38992089-504726f8-43fd-11e8-915a-d02bd305ad46.PNG)

which have 16 core CPU and other is P2.xlarge GPU compute having 4 core of CPU. In First case CPU utilisation is very less I am not understanding where it is busy since the instance has only model to run

**Comment by AlexeyAB** [comment]
Hi,

https://aws.amazon.com/ec2/instance-types/

* **P2.xlarge** - has GPU NVIDIA K80 and CPU Intel Xeon E5-2686 v4 (Broadwell) **(4 Cores)**
* **m5.4xlarge** - doesn't have GPU, but has CPU Intel Xeon® Platinum 8175 AVX-512 **(16 Cores)**

Did you compile Darknet for CPU or for GPU in each of 2 cases?

**Comment by deepkshikha** [comment]
I am training my model seperately in both case yes I compiled in each case seperately

Original one

**Comment by AlexeyAB** [comment]
The graphs of the CPU load are shown correctly

* **P2.xlarge** - uses **GPU**, so CPU is almost unused
* **m5.4xlarge** - uses CPU, so CPU is almost used

**Comment by deepkshikha** [comment]
But it shows only 20 % utilization for m5.x4large case it should be 80-100
since only CPU is getting utilised??

**Comment by AlexeyAB** [comment]
Probably this repo https://github.com/KleinYuan/easy-yolo doesn't use multithreading for convolution (GEMM).
Use this GitHub repository with `OPENCV=1 AVX=1` in the Makefile on m5.4xlarge

**Comment by deepkshikha** [comment]
@AlexeyAB I tried I am getting below error in your repository


darknet: ./src/utils.c:210: error: Assertion `0' failed.
Aborted (core dumped)

**Comment by deepkshikha** [comment]
One more thing how can I calculte True Negative I want to draw TPR Vs FPR curve

**Comment by AlexeyAB** [comment]
Do you use classifier or detector?

**Comment by deepkshikha** [comment]
Detector

**Comment by AlexeyAB** [comment]
@deepkshikha 

On Linux you should compile with `OPENMP=1 AVX=1` in the Makefile to use multi-threading (to utilize all cores of CPU): https://github.com/AlexeyAB/darknet/blob/e18a303d1c17acd30c21162826c554ee770cf35a/Makefile#L4-L5

**Comment by deepkshikha** [comment]
# error "SSE4A instruction set not enabled"
   ^
Itried I am getting this

**Comment by AlexeyAB** [comment]
On what Instance this error doesn occur? Use `AVX=0`


It shouldn't occur on the instance **m5.4xlarge** - that  has CPU Intel Xeon® Platinum 8175 **AVX-512** (16 Cores)

**Comment by AlexeyAB** [comment]
> which one is for multithreading ?? I am getting this on m5.4xlarge for AVX=1

* `OPENMP=1` for multi-threading on CPU
* `AVX=1` for vectorization on CPU

----

Try change this line in the Makefile: https://github.com/AlexeyAB/darknet/blob/a2f3644e87c11c222f86aa2aeb0e31d858ebb024/Makefile#L60
to this `CFLAGS+= -ffp-contract=fast -mavx -msse4.1 -msse4a`

Use `OPENMP=1 AVX=1` in the Makefile and recompile. 
Will this error occur on m5.4xlarge after that?

**Comment by deepkshikha** [comment]
Thanks a  @AlexeyAB

**Comment by AlexeyAB** [comment]
@deepkshikha 

> `CFLAGS+= -ffp-contract=fast -mavx -msse4.1 -msse4a`

Did this fix help, and does `OPENMP=1 AVX=1` work well on m5.4xlarge now?

**Comment by deepkshikha** [comment]
Yes It worked Thank you

**Comment by deepkshikha** [comment]
@AlexeyAB It's running but no weight is appearing in backup folder

**Comment by deepkshikha** [comment]
@AlexeyAB  But no weight is appearing in the backup. Although its running and using 100% utilization

**Comment by deepkshikha** [comment]
@AlexeyAB But no weight is appearing in the backup. Although its running and using 100% utilization

**Comment by AlexeyAB** [comment]
How many iterations did you train?
weights-file will be saved for each 100 iterations 

* On CPU `m5.4xlarge` it take a long time
* But training ~20x times faster on `p2.xlarge ` with `GPU=1 CUDNN=1`.
* And training ~100x times faster on `p3.2xlarge` with `GPU=1 CUDNN=1 CUDNN_HALF=1`.

----

https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> 9. After training is complete - get result yolo-obj_final.weights from path build\darknet\x64\backup\
> 
> After each 100 iterations you can stop and later start training from this point. For example, after 2000 iterations you can stop training, and later just copy yolo-obj_2000.weights from build\darknet\x64\backup\ to build\darknet\x64\ and start training using: darknet.exe detector train data/obj.data yolo-obj.cfg yolo-obj_2000.weights
> 
> (in the original repository https://github.com/pjreddie/darknet the weights-file is saved only once every 10 000 iterations if(iterations > 1000))

**Comment by deepkshikha** [comment]
Yes on p2.xlarge its quicker I wanna check actually If we increase number of CPU core how much speed I can gain . Can I run this on Amazon Lambda to reduce GPU cost ??

**Comment by AlexeyAB** [comment]
* If you use `m5.4xlarge` with `OPENMP=1 AVX=1` then speed propotional to the number of CPU cores

* If you use `p2.xlarge` or `p3.2xlarge` with `GPU=1 CUDNN=1 OPENCV=1` then increasing of CPU cores **willn't give you any speedup**

* If you use `p2.xlarge` or `p3.2xlarge` with `GPU=1 CUDNN=1 OPENCV=0` then CPU cores should be not less than 4 otherwise data-autgmentation can be a bottleneck

**Comment by AlexeyAB** [comment]
> Can I run this on AWS Lambda to reduce GPU cost ??

As I know the GPU is not available at this time in the AWS Labmda. 
So AWS Labmda will not be cheaper than using `p3.2xlarge` with `GPU=1 CUDNN=1 OPENCV=1 CUDNN_HALF=1` even if you use `p3.2xlarge` only 1/10 of the paid time.


https://forums.aws.amazon.com/thread.jspa?messageID=630312

> Posted on: Feb 9, 2018 2:41 PM
> Bump. Any update on this? Just found out about Lambda, but if there's no GPU access I can't use it.

But in general I think you can use Darknet CPU-only on AWS Lambda, just link C-code with Java/Python or do C-addon for Node.js.

**Comment by deepkshikha** [comment]
Thanks a lot

**Comment by deepkshikha** [comment]
How have you computed 1/10 because there will be memory latency and other stuff So Is there any formula?

**Comment by AlexeyAB** [comment]
This is very aproximately. You should find info about performance GFlops Single-precision/Mixed-precision for: 

* GPU Tesla V100 on `p3.2xlarge` **(~100 000 Gflops FP16/32)** / 3$ ~= **33 000**

* GPU NVIDIA K80 on `p2.xlarge`  **(~8 000 Gflops FP32)** / 0.9$ ~= **9 000**

* CPU Intel Xeon® Platinum 8175 on `m5.4xlarge` **(~2 000 Gflops FP32)** / 0.77$ ~= **2600**

Compare relation `Gflops / Price$`:
* `p2.xlarge` is **3.5 x times** better than `m5.4xlarge`
* `p3.2xlarge` is **13 x times** better than `m5.4xlarge`

Then find info about `Gflops / Price$` for AWS Labmda.
